src/popup/lib/utils.ts

```diff
@@ -29,3 +29,29 @@ export function roundWithPrecision(num: number, precision: number) {
   const multiplier = Math.pow(10, precision)
   return Math.round(num * multiplier) / multiplier
 }
+
+export function formatNumber(
+  value: number,
+  scale: number,
+  allowExponential = false
+): string {
+  if (
+    scale <= 2 ||
+    (value * 100 >= 1 && value * 100 - Math.floor(value * 100) === 0)
+  ) {
+    return value.toFixed(2).toString()
+  } else if (scale >= 3 && scale <= 4) {
+    return value.toString()
+  } else {
+    if (value * 10 ** 4 >= 1 || !allowExponential) {
+      let fixedScale = 5
+      while (
+        value * 10 ** fixedScale - Math.floor(value * 10 ** fixedScale) > 0 &&
+        fixedScale < scale
+      ) {
+        ++fixedScale
+      }
+      return value.toFixed(fixedScale).toString()
+    } else return value.toExponential()
+  }
```

Comment:
I think we can simplify the function. Let me know what you think about this approach:

```suggestion
export function formatNumber(
  value: number,
  scale: number,
  allowExponential = false
): string {
  // Check for NaN as well
  if(Number.isNan(value)) return "0.00"
  if (
    scale <= 2 ||
    (value * 100 >= 1 && value * 100 - Math.floor(value * 100) === 0)
  ) {
    return value.toFixed(2)
  }
  // Call `toFixed` with the asset's scale and remove trailing zeros
  return allowExponential ? value.toExponential() : value.toFixed(scale).replace(/0+$/, '')
}
```